### PR TITLE
1.21.3 update

### LINKED ISSUE #1702: New input packet
### Feature description

Latest snapshot 24w36a just added a keyboard input packet.
```java
Client:
@Override
public void tick() {
if (this.getWorld().isPosLoaded(this.getBlockX(), this.getBlockZ())) {
	super.tick();
	this.sendSneakingPacket();
	if (!this.field_54156.equals(this.input.field_54155)) {
		this.networkHandler.sendPacket(new PlayerInputC2SPacket(this.input.field_54155));
		this.field_54156 = this.input.field_54155;
	}
```
```java
@Override
	public void onPlayerInput(PlayerInputC2SPacket packet) {
		NetworkThreadUtils.forceMainThread(packet, this, this.player.getServerWorld());
		this.player.method_63561(packet.input());
	}
```

```java
	public record class_10185(boolean forward, boolean backward, boolean left, boolean right, boolean jump, boolean shift, boolean sprint) {
```

This opens up the possibility for a huge amount of cool new checks.

### LINKED ISSUE #1769: BADPACKETSE false for 1.21.2+ clients
### Describe the false positive and how to replicate it

Ride any vehicle in 1.21.2 (using viaversion, not a viaversion issue though, I think they made clients send flying packets while riding in 1.21.2+, or changed how steer vehicle packets are sent, idk)

### Grim version

b0cd32622e0dbaa1d6bded568e4db80d1ec21d6d a bit out of date but im pretty sure same issue happens on latest grim

### Server version

This server is running Paper version 1.21.1-DEV-master@14a48cd (2024-10-21T13:11:50Z) (Implementing API version 1.21.1-R0.1-SNAPSHOT)
You are 1 version(s) behind
Download the new version at: https://papermc.io/downloads/paper
Previous version: 1.21.1-DEV-7cd4f2c (MC: 1.21.1)

### Client version

1.21.3

### Plugins

Grim

### PR DESCRIPTION
**Pull request by SamB440**
- Added support for 1.21.3 vehicles
- Added support for new end tick packet
- Added support for new player input packet
- Fixed precision loss when jumping and/or in the air (we are now 1e-15 average)

~~For some reason, sneaking whilst sprinting falses on 1.21.3, yet the movement is the same as older versions. Someone needs to figure out the issue there.~~

ViaVersion setups are not supported and will not be supported until we have the ability to bypass via for certain listeners.

Fixes #1769 
Closes #1702

### CODE REVIEW

**Comment by ewof** [comment]
```
[01:19:11 WARN]: [GrimAC] Global task for GrimAC v2.3.68 generated an exception
java.lang.IllegalStateException: Async chunk retrieval
        at ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(TickThread.java:52) ~[parchment-1.21.3.jar:1.21.3-DEV-0249f75]
        at org.bukkit.craftbukkit.CraftWorld.getChunkAt(CraftWorld.java:371) ~[parchment-1.21.3.jar:1.21.3-DEV-0249f75]
        at grimac-2.3.68.jar/ac.grim.grimac.events.packets.patch.ResyncWorldUtil.lambda$resyncPositions$2(ResyncWorldUtil.java:78) ~[grimac-2.3.68.jar:?]
        at io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler.lambda$execute$0(FoliaGlobalRegionScheduler.java:47) ~[parchment-1.21.3.jar:?]
        at io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler$GlobalScheduledTask.run(FoliaGlobalRegionScheduler.java:179) ~[parchment-1.21.3.jar:?]
        at io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler.tick(FoliaGlobalRegionScheduler.java:37) ~[parchment-1.21.3.jar:?]
        ```

**Comment by Cyramek** [comment]
fastbreak should also be updated?

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java [hunk 1]
```diff
@@ -99,6 +100,33 @@ public void onPacketReceive(PacketReceiveEvent event) {
...
             player.packetStateData.receivedSteerVehicle = true;
+        } else if (event.getPacketType() == PacketType.Play.Client.PLAYER_INPUT) {
+            GrimPlayer player = GrimAPI.INSTANCE.getPlayerDataManager().getPlayer(event.getUser());
+            if (player == null) return;
+
+            WrapperPlayClientPlayerInput input = new WrapperPlayClientPlayerInput(event);
+            byte forward = 0;
+            byte sideways = 0;
+            if (input.isForward()) {
+                forward += 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                forward++;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java [hunk 2]
```diff
@@ -99,6 +100,33 @@ public void onPacketReceive(PacketReceiveEvent event) {
...
+            }
+
+            if (input.isBackward()) {
+                forward -= 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                forward--;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java [hunk 3]
```diff
@@ -99,6 +100,33 @@ public void onPacketReceive(PacketReceiveEvent event) {
...
+            }
+
+            if (input.isLeft()) {
+                sideways += 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                sideways++;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java [hunk 4]
```diff
@@ -99,6 +100,33 @@ public void onPacketReceive(PacketReceiveEvent event) {
...
+            }
+
+            if (input.isRight()) {
+                sideways -= 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                sideways--;
```

File: src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java [hunk 5]
```diff
@@ -700,7 +703,36 @@ private void loopVectors(GrimPlayer player, Set<VectorData> possibleVectors, flo
...
+        int strafeMax = 1;
+
+        // Calculate inputs by the players known inputs on 1.21.2+
+        if (player.supportsEndTick()) {
+            forwardMin = forwardMax = strafeMin = strafeMax = 0;
+
+            final KnownInput knownInput = player.packetStateData.knownInput;
+            if (knownInput.forward()) {
+                forwardMax += 1;
+                forwardMin += 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java]:
```suggestion
                forwardMax++;
                forwardMin++;
```

File: src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java [hunk 6]
```diff
@@ -700,7 +703,36 @@ private void loopVectors(GrimPlayer player, Set<VectorData> possibleVectors, flo
...
+            }
+
+            if (knownInput.backward() && (!player.isSprinting || player.isSwimming)) {
+                forwardMax -= 1;
+                forwardMin -= 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java]:
```suggestion
                forwardMax--;
                forwardMin--;
```

File: src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java [hunk 7]
```diff
@@ -700,7 +703,36 @@ private void loopVectors(GrimPlayer player, Set<VectorData> possibleVectors, flo
...
+            }
+
+            if (knownInput.left()) {
+                strafeMax += 1;
+                strafeMin += 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java]:
```suggestion
                strafeMax++;
                strafeMin++;
```

File: src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java [hunk 8]
```diff
@@ -700,7 +703,36 @@ private void loopVectors(GrimPlayer player, Set<VectorData> possibleVectors, flo
...
+            }
+
+            if (knownInput.right()) {
+                strafeMax -= 1;
+                strafeMin -= 1;
```
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/predictionengine/predictions/PredictionEngine.java]:
```suggestion
                strafeMax--;
                strafeMin--;
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java (hunk 1 above)
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                forward++;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java (hunk 2 above)
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                forward--;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java (hunk 3 above)
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                sideways++;
```

File: src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java (hunk 4 above)
Comment (ManInMyVan) [inline src/main/java/ac/grim/grimac/events/packets/PacketPlayerSteer.java]:
```suggestion
                sideways--;
```

**Review by ManInMyVan — CHANGES_REQUESTED** [review]

**Comment by SamB440** [comment]
> fastbreak should also be updated?

What should be?

**Comment by ManInMyVan** [comment]
> > fastbreak should also be updated?
> 
> What should be?

instead of being based on ms, it should be based on ticks, like I did for pre 1.9 clients in my breaking checks pr

**Comment by SamB440** [comment]
I don't see anything 1.8-like / specific in fast break. There's only `On 1.8 clients, via screws with this packet meaning we must fall back to the 1.8 idle flying packet`, but we're _meant_ to be using animation packet

**Pull request merged**